ipykernel/ipkernel.py

```diff
@@ -348,25 +342,23 @@ def do_apply(self, content, bufs, msg_id, reply_metadata):
                 item_threshold=self.session.item_threshold,
             )
 
-        except:
+        except BaseException as e:
             # invoke IPython traceback formatting
             shell.showtraceback()
-            # FIXME - fish exception info out of shell, possibly left there by
-            # run_code.  We'll need to clean up this logic later.
-            reply_content = {}
-            if shell._reply_content is not None:
-                reply_content.update(shell._reply_content)
-                # reset after use
-                shell._reply_content = None
-
-                # FIXME: deprecate piece for ipyparallel:
-                e_info = dict(engine_uuid=self.ident, engine_id=self.int_id, method='apply')
-                reply_content['engine_info'] = e_info
+            reply_content = {
+                u'traceback': shell._last_traceback or [],
+                u'ename': unicode_type(type(e).__name__),
+                u'evalue': safe_unicode(e),
+            }
+            # FIXME: deprecate piece for ipyparallel:
+            e_info = dict(engine_uuid=self.ident, engine_id=self.int_id, method='apply')
+            reply_content['engine_info'] = e_info
```

Comment:
@minrk this is in the handler for `apply` messages for ipyparallel - can you check if it makes sense?